plugins/flytekit-pydantic/flytekitplugins/pydantic/basemodel_extensions.py

```diff
@@ -0,0 +1,152 @@
+import abc
+from typing import Any, Dict, Generic, Optional, Type, TypeVar, Union, cast
+
+from typing_extensions import TypedDict
+
+from flytekit.types.directory import types as flyte_directory_types
+from flytekit.types.file import file as flyte_file
+
+from . import flytepath_creation
+
+## ====================================================================================
+# Base class
+## ====================================================================================
+
+
+Serializable = TypeVar("Serializable")
+SerializedFlyteType = TypeVar("SerializedFlyteType")
+
+
+def get_pydantic_flyteobject_config() -> Dict[str, Any]:
+    """
+    Returns the pydantic config for the serializers/deserializers
+    """
+    return {
+        "json_encoders": {
+            serializer_deserializer()._t: serializer_deserializer().serialize  # type: ignore
+            for serializer_deserializer in PydanticSerializerDeserializerBase.__subclasses__()
+        },
+    }
+
+
+class PydanticSerializerDeserializerBase(
+    abc.ABC,
+    Generic[Serializable, SerializedFlyteType],
+):
+    """Base class for object serializers/deserializers"""
+
+    @abc.abstractmethod
+    def serialize(self, obj: Serializable) -> SerializedFlyteType:
+        pass
+
+    @abc.abstractmethod
+    def deserialize(self, obj: SerializedFlyteType) -> Serializable:
+        pass
+
+    def __init__(self, t: Type[Serializable]):
+        self._t = t
+        self._set_validator_on_serializeable(t)
+
+    def _set_validator_on_serializeable(self, serializeable: Type[Serializable]) -> None:
+        """
+        Sets the validator on the pydantic model for the
+        type that is being serialized/deserialized
+        """
+        setattr(serializeable, "__get_validators__", lambda *_: (self.deserialize,))
+
+
+## ====================================================================================
+#  FlyteDir
+## ====================================================================================
+
+
+class FlyteDirJsonEncoded(TypedDict):
+    """JSON representation of a FlyteDirectory"""
+    path: str 
+
+
+FLYTEDIR_DESERIALIZABLE_TYPES = Union[flyte_directory_types.FlyteDirectory, str, FlyteDirJsonEncoded]
+
+
+class FlyteDirSerializerDeserializer(
+    PydanticSerializerDeserializerBase[flyte_directory_types.FlyteDirectory, FlyteDirJsonEncoded]
+):
+    def __init__(
+        self,
+        t: Type[flyte_directory_types.FlyteDirectory] = flyte_directory_types.FlyteDirectory,
+    ):
+        super().__init__(t)
+
+    def serialize(self, obj: flyte_directory_types.FlyteDirectory) -> FlyteDirJsonEncoded:
+        return {"path": obj.remote_source if obj.remote_source else obj.path}
```

Comment:
should we upload the local file to s3? like https://github.com/flyteorg/flytekit/blob/master/flytekit/types/file/file.py#L393-L397

Reply:
Latest commit should cover that - how does this look?